aip/0146.md

```diff
@@ -0,0 +1,91 @@
+introduce the "least generic" approach that is able to satisfy the use case.
+
+### Oneof
+
+A `oneof` **may** be used to introduce a type union: the user or service is
+able to specify one of the fields inside the oneof. Additionally, a `oneof`
+**may** be used with the same type (usually strings) to represent a semantic
+difference between the options.
+
+Because the individual fields in the oneof have different keys, a developer can
+programatically determine which (if any) of the fields is populated.
+
+A `oneof` preserves the largest degree of type safety and semantic meaning for
+each option, and services **should** generally prefer them over other generic
+or polymorphic options when feasible. However, the `oneof` construct is
+ill-suited when there is a large (or unlimited) number of potential options, or
+when there is a large resource structure that would require a long series of
+"cascading oneofs".
+
+**Note:** Adding additional possible fields to an existing `oneof` is a
+non-breaking change, but moving existing fields into or out of a `oneof` is
+breaking (it creates a backwards-incompatible change in Go protobuf stubs).
+
+### Maps
+
+Maps **may** be used in situations where a situation where many values _of the
+same type_ are needed, but the keys are unknown or user-determined.
+
+Maps are usually not appropriate for generic fields because the map values all
+share a type, but occasionally they are useful. In particular, a map can
+sometimes be suited to a situation where many objects of the same type are
+needed, with different behavior based on the names of their keys (for example,
+using keys as environment names).
+
+### Struct
+
+The [`google.protobuf.Struct`][] object **may** be used to represent arbitrary
+nested JSON. Keys can be strings, and values can be floats, strings, booleans,
+arrays, or additional nested structs, allowing for an arbitrarily nested
+structure that can be represented as JSON (and is automatically represented as
+JSON when using REST/JSON).
+
+A `Struct` is most useful when the service does not know the schema in advance,
+or when a service needs to store and retrieve arbitrary but structured user
+data. Using a `Struct` is convenient for users in this case because they can
+easily get JSON objects that can be natively manipulated in their environment
+of choice.
+
+If a service needs to reason about the _schema_ of a `Struct`, it **should**
+use [JSONSchema][] for this purpose. Because JSONSchema is itself JSON, a valid
+JSONSchema document can itself be stored in a `Struct`.
+
+### Any
+
+The [`google.protobuf.Any`][] object can be used to send an arbitrary
+serialized protocol buffer and a type definition. However, this introduces
+complexity, because the consumer of an `Any` needs to register the type and go
+through a deserialization process before being able to get a usable object.
+Because of this, `Any` **should not** be used unless other options are
+infeasible.
```

Comment:
One interesting point which may be worth mentioning is that the JSON representation of an Any is compatible with a Struct.
That's useful because it means data can be transported efficiently between systems that *do* know all the protobuf types involved, and then serialized to JSON across a boundary, then deserialized as a `Struct` for systems that might not know the protobuf types involved.

The big difference between Any and Struct is that Any is effectively useless for anything other than data propagation unless the code using it has access to the proto.

That was a somewhat incoherent attempt at explaining what I mean - happy to VC about it instead.

Reply:
Actually, I know precisely what you mean. I have changed this somewhat to incorporate this.

I am not mentioning the Any/Struct point you made, which tends to be untrue across public API surfaces (at least when using gRPC).